samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -461,8 +483,25 @@ public void onJobModelExpired() {
             } else {
               LOGGER.info("Container: {} shutdown completed for stream processor: {}.", container, processorId);
             }
+          } else if (state == State.IN_REBALANCE) {
+            if (container != null) {
+              boolean hasContainerShutdown = interruptSamzaContainer();
```

Comment:
It appears that upon a successful interrupt on a SamzaContainer, it is required to create a new `containerExecutorService`. Since the lifecycle of the `containerExecutorService` is dependent on the result of `interruptSamzaContainer()` it might be better to pass in the `containerExecutorService` as a parameter to the method to make that more obvious to the caller in case this method needs to be used elsewhere in the future.

Reply:
I agree that, it shuts down the executor service and the call site can instantiate a new one potentially based on its result. It currently shuts it down or not based on `container != null` and maybe i will extract that out from the method and establish a clear contract on this method where it will always invoke shutdownNow. By that, it is clear on the call site that upon invocation of this method, if executor service needs to used a new one needs to be created. 

I prefer to use immutable method parameters wherever possible since its easier to reason about and also not worry about the side effects. In this case, this method is instance method anyways and IMO, I don't quite understand the benefit of passing the containerExecutorService as a method parameter.

As far as creating an executor service upon success, I am debating whether to do it here vs create a new executor service in `onNewJobModel`. wdyt?

Reply:
I think here is fine since this is where it shuts down.